### ISSUE TITLE
Ability to Add onsubmit attribute to Mautic Forms.

### ISSUE THREAD

**Issue by sambarnes90** [mautic/mautic#2093]
What type of report is this:

| Q | A |
| --- | --- |
| Bug report? | N |
| Feature request? | Y |
| Enhancement? | N |
## Description:

Currently we can add custom attributes to fields in any Mautic form.

However, we cannot add custom attributes to the form itself. We may be able to add this to the form's template, but adding it on an individual form by form basis is not possible at the moment.

Primarily this would be to add onsubmit events to the form's HTML.

For example:

`<form autocomplete="false" role="form" method="post" action="https://mautic.example.com/form/submit?formId=111" id="mauticform_example1" data-mautic-form="example1" target="mauticiframe_example1" onsubmit="somefunction();"></form>
`

**Comment by CreaoticX** [comment]
I don't think this feature is actually what you want. Mautic already automatically adds the onsubmit function to the form.  You can see this in https://github.com/mautic/mautic/blob/staging/media/js/mautic-form-src.js#L144 .  However, if you add an onsubmit to the form Mautic doesn't add the default functionality to the form.

So, I'm assuming you're trying to do one of two things.  Either 1. you want to execute some code when the form is submitted.  In which case, what you really want to do is create and add a custom callback to probably the `onValidateEnd` event (https://github.com/mautic/mautic/blob/staging/media/js/mautic-form-src.js#L306) because then you can tell if the form was valid before you execute your code.  Or  2. you're trying to override the default form validation and submission that comes with Mautic.  In which case I'd say you should use a custom theme, because adding an onsubmit option to forms would cause any form that includes it to break.  I think this would be counterintuitive for most users and would cause more problems than benefits.

Does that make sense or do you have some alternative reasoning behind adding the onsubmit option?

**Comment by alanhartless** [comment]
Ugh - I'm sorry guys. I really really need to get this documented. I'll make add it to my todo list to do over the weekend. 

```
var MauticFormCallback = {
    replaceWithFormName: {
        onValidateEnd: function () {
             // before form submit
        },
        onResponse: function (response) { 
             // after form submit
        }
    }
};```

**Comment by alanhartless** [comment]
Sorry - I'l reopen this in order to add support for other attributes - although we'll likely not allow an onsubmit or else it'll break internal workings.

**Comment by CreaoticX** [comment]
@alanhartless  Hey, thanks for supplying the code.  I knew something like that was needed, but I didn't know the code myself.

**Comment by hmaesta** [comment]
Hey guys. 
On `onValidateEnd`, how can I change some data before send?

https://www.mautic.org/community/index.php/8664-handle-form-data-before-send

**Comment by weismannweb** [comment]
Is this only for server side in the Mautic side within the iframe?  I could not get the callbacks to work client side on the website the form code was placed.  I may be missing something but anyway I created this to do it on the website side for different conversions we needed:

`<script type="text/javascript" src="https://cdn.rawgit.com/meetselva/attrchange/master/js/attrchange.js"></script>`

```
$( document ).ready(function() {
  
  $('.mautic-conversion').map(function() {
    mformname = this.getAttribute('data-mautic-form');
  mdivid = 'mauticform_wrapper_'+mformname;
    
    console.log('mautic: #'+mdivid);
    
  jQuery('#'+mdivid).attrchange({
  trackValues: true, 
  callback: function (event) {
    console.log('mautic: '+event.attributeName);
    if(jQuery('#'+mdivid).hasClass( "mauticform-post-success")){
    	console.log('convert');
      	gpc_conversion('Mautic Form','Submit',mformname);
    	}
      }
  });
    
    

});
 });
```
If you want a form to do a conversion then in mautic give it that class.

Long story short I noticed that the class on the div above the form added .mauticform-post-success to the class attribute.  So I used a jquery library that detects mutations ie changes to the attributes of elements.

Anyway I hope its useful and if it was the wrong thing to do and there IS a way to use these callbacks on the website without touching Mautic code please someone let me know how?

**Comment by pierrejoye** [comment]
For other as this ticket is found in many similar questions:

https://www.leuchtfeuer.com/en/mautic-know-how/mautic/dynamic-values-in-mautic-forms-date-time-url/

That way the hooks will be "injected" at the same time than the form and works smoothly for any external site.

All possible hooks are very well documented in the documentation here <3:

https://developer.mautic.org/#js-form-processing-hooks
